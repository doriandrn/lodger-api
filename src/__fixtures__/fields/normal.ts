const fields = {
  x1: { },
  x2: { required: true, index: true },
  x3: { type: 'number' }
}

export default fields
