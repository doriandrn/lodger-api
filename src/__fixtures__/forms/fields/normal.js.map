{"version":3,"file":"normal.js","sourceRoot":"","sources":["normal.ts"],"names":[],"mappings":"AAAA,MAAM,MAAM,GAAG;IACb,EAAE,EAAE,EAAE,IAAI,EAAE;IACZ,EAAE,EAAE,EAAE,IAAI,EAAE,QAAQ,EAAE,IAAI,EAAE,KAAK,EAAE,IAAI,EAAE;IACzC,EAAE,EAAE,EAAE,IAAI,EAAE;CACb,CAAA;AACD,eAAe,MAAM,CAAA"}