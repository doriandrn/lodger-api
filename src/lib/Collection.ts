import Schema, { LodgerSchemaCreator } from './Schema'
import { RxCollectionCreator } from 'rxdb'

class Collection<Name extends string, Interface> extends Schema<Name, Interface> implements RxCollectionCreator {
  constructor (data: LodgerSchemaCreator<Interface>) {
    super(data)
  }
}
