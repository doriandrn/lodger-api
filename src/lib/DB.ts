import * as RxDB from 'rxdb'

import memoryAdapter from 'pouchdb-adapter-memory'
import idbAdapter from 'pouchdb-adapter-idb'
import httpAdapter from 'pouchdb-adapter-http'

import { env } from 'defs/env'

// RxDB.QueryChangeDetector.enable()
// RxDB.QueryChangeDetector.enableDebugging()

// const adapters = {
//   production: ['http', 'idb'],
//   development: ['memory']
// }

// Object.keys(adapters).forEach(env, () => {
//   adapters[env].forEach((adapterType: Adapter) => {
//     if (NODE_ENV !== env) return
//     const adapter = `pouchdb-adapter-${adapterType}`
//     RxDB.plugin(adapter)
//   })
// })

switch (env) {
  default:
    RxDB.plugin(memoryAdapter)
    break

  case 'production':
    RxDB.plugin(httpAdapter)
    RxDB.plugin(idbAdapter)
    break
}

export default RxDB

// /**
//  * RxDB instantiator
//  *
//  * @export
//  * @async
//  * @param {RxDB.RxCollectionCreator[]} collections
//  * @param {RxDB.RxDatabaseCreator} [config]
//  * @returns {RxDB} the fresh database
//  */
// export default async function (
//   collections: RxDB.RxCollectionCreator[],
//   config?: RxDB.RxDatabaseCreator
// ) {
//   debug('Initing')
//   const db =

//   // show leadership in title
//   db.waitForLeadership().then(() => {
//     if (env !== 'dev') return
//     if (!process.browser && env !== 'dev') return
//     document.title = `â™› ${document.title}`
//   })
//   await Promise.all(collections.map(c => db.collection(c)))
//   return db
// }
