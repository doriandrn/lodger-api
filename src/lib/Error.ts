// import fx from 'fx'

/**
 * Error logger
 */
export default class LodgerError extends Error {
  constructor(m: string, details?: any) {
    if (details) {
      m = m.replace('%%', `"${JSON.stringify(details)}"`)
    }
    super(m)

    // Set the prototype explicitly.
    Object.setPrototypeOf(this, LodgerError.prototype)
  }
}
