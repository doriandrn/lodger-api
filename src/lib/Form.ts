/**
 * Forms for Lodger
 * are quite diferrently structured
 * than a normal JsonSchema
 */
import Debug from 'debug'
import { RxCollectionCreator } from 'rxdb'

import { env } from './defs/env'
import FormError from './Error'
import { GetterTree } from 'vuex'
import { RootState } from './Store'

type ItemReference = Plural<Taxonomie> | object
type FormExcludables = 'db' | 'addForm' | 'editForm' | 'all'
type ItemExcludableFrom = FormExcludables[]

export type FieldCreator = {
  id: string,
  name?: string,

  label?: string
  placeholder?: string

  type?: FormItemTypes
  required?: boolean
  encrypted?: boolean

  default?: any
  value?: any

  step?: number,
  index?: boolean,
  ref?: ItemReference
  items?: object
  indexRef?: boolean

  excludeFrom?: ItemExcludableFrom

  v?: string // validation string
  click?: string
  showInList?: 'primary' | 'secondary' | 'details'[]
}



/**
 * Errors Definition
 * @readonly
 * @enum {string}
 *
 * @todo account for translations
 */
enum Errors {
  invalidRequested = 'Invalid form requested: %%',
  invalidName = 'Invalid name supplied',
  noData = 'Form %% is missing data',
  missingName = 'Forms should have a name',
  missingPlural = 'A plural definition is required for %%'
}

if (env === 'test')
  Debug.enable('Form:*')

export type LodgerFormCreator = {
  name?: string
  plural: Plural<Taxonomie>
  fields: FieldCreator[]

  methods?: { [k: string]: () => void }
  statics?: { [k: string]: () => void }
  sync?: boolean
  setari?: any
}

const formsPath = ['dev', 'test'].indexOf(env) > -1 ? 'forms' : '.'


// export interface LodgerFormConstructor {
//   new (data: LodgerFormCreator): LodgerForm
// }

interface LodgerForm {
  name: string
  collection: undefined | RxCollectionCreator
  indexables ?: string[]
}

/**
 * Forms are read from within the `lib/forms/` directory
 *
 * @class Form
 * @implements {LodgerForm}
 */
class Form implements LodgerForm {
  name: string
  fields: FieldCreator[]
  collection: undefined | RxCollectionCreator

  readonly indexables ?: string[]
  readonly plural : Plural<Taxonomie>

  /**
   * Creates an instance of Form.
   *
   * @param {LodgerFormCreator} data - Form input data
   * @param {boolean} [generateRxCollection=true] - some forms don't require this
   *
   * @memberof Form
   */
  constructor (
    data: LodgerFormCreator,
    generateRxCollection : boolean = true
  ) {
    const { fields, name, plural, methods, statics } = data
    if (!name) throw new FormError('Form should have a name %%', data)
    if (!fields || !Object.keys(fields).length)
      throw new FormError('missing fields on form %%', name)

    this.name = name
    this.plural = plural
    this.fields = fields

    if (generateRxCollection) {
      const schema = new Schema(data, true)
      const collection = {
        name: plural,
        schema,
        methods,
        statics
      }
      this.collection = collection
      this.indexables = Object.keys(schema.properties).filter(prop => schema.properties[prop].index)
    }
  }

  /**
   * Makes a Vue-ready $data {object} suitable to be completed
   * by the user in the end form
   * as it will turn reactive
   *
   * @summary for new forms, values are all undefined
   *
   * @param {Boolean} isNewForm - mostly used for 'add' forms
   * @returns {Object} data item (Vue $data - ready)
   */
  value (
    isNewForm: boolean,
    getters?: GetterTree<any, RootState>
  ): FormValue {
    const { fields, name } = this
    // const debug = Debug('Form:value')
    let $data = {} as any

    fields.forEach(camp => {
      const { label, required, click, excludeFrom } = camp
      let { id, value } = camp

      let _def = camp.default

      if (click && !id) camp.id = click

      // skip excluded fields
      if (isNewForm && excludeFrom && (excludeFrom.indexOf('db') || excludeFrom.indexOf('addForm')))
        value = undefined

      // apply getters to funcs
      value = typeof value === 'function' && getters ? value(getters) : undefined
      _def = typeof _def === 'function' ? _def(getters) : undefined

      // label
      camp.label = label || `${name ? `${name}.new.` : ''}${id}`

      // validarea de required
      if (required || (camp.v && camp.v.indexOf('required') < 0))
        camp.v = `required|${camp.v || ''}`

      // valoarea finala
      $data[id] = null
      $data[id] = value !== null && value !== undefined ? value : _def
    })

    return this.handleOnSubmit($data)
  }

  /**
   * Manipulates the final data before submitting the form to the DB
   *
   * @param data
   */
  handleOnSubmit (
    data : FormValue
  ) {
    const manipulatedData: any = {}

    // not data.denumire pt servicii :/
    if (!data.la && !data.denumire) data.la = Date.now()
    Object.keys(data).forEach(what => {
      let value = data[what]
      if (value === null || value === 'undefined') {
        return
      }

      manipulatedData[what] = value
    })

    if (!context) return manipulatedData
    const { referencesIds } = context

    Object.assign(manipulatedData, referencesIds)

    return manipulatedData
  }

  /**
   * Loads a 'known' form by name
   *
   * @param name
   */
  static async load (name: string): Promise<Form> {
    const debug = Debug('lodger:Form')
    if (!name)
      throw new FormError('no name supplied for form')

    const formPath: string = `${formsPath}/${String(name).toLowerCase()}`

    try {
      const formData: LodgerFormCreator = await import(formPath)
      Object.assign(formData, { name })
      debug('âœ“', name)
      return new Form(formData)
    } catch (err) {
      debug('x', name)
      throw new FormError(err)
    }

  }
}

export {
  Form,
  Errors,
  prepareRxSchema
}
