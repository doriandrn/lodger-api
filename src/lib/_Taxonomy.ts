import { predefinite } from 'forms/serviciu'


// /**
//  * Pt taxonomia ceruta
//  * ia formul
//  * si tot ce are nevoie de Id de altceva
//  * se populeaza
//  *
//  * @param {Object} { references, getters }
//  * @returns {Object} eg { asociatieId: 'XXXX' }
//  */
// function assignRefIdsFromStore (context: any) {
//   const { references, getters } = context
//   if (!(references && references.length)) return

//   const refsObj = {}

//   references.map((tax: Taxonomie) => {
//     refsObj[`${tax}Id`] = getters[`${tax}/selected`]
//   })

//   return refsObj
// }


// const xx: LodgerTaxonomyCreator<Taxonomie> = new Taxonomy('asociatie', 'form', 'collection')
/**
 * gets the sorting options for tax
 * @returns an object with each key used as a sorting option
 */
// get sortOptions () {
//   const { indexables, name } = this

//   if (!['serviciu', 'contor'].indexOf(name)) {
//     indexables.push('la')
//   }

//   // TODO: !!! ia din common methods
//   const sorts = {}
//   indexables.forEach(indexable => {
//     const label = `sort.${indexable === 'name' ? 'az' : indexable}`
//     Object.assign(sorts, { [indexable]: { label } })
//   })

//   // debug(`${name} => sortable fields`, sorts)

//   return sorts
// }
/**
 * Items to be display to user,
 * @returns {Object} the keys of the fields: their position
 *
 */
// get __displayItemKeys () {
//   const { fields } = this.data

//   return Object.assign({}, ...fields
//     .filter(field => field.showInList)
//     .map(field => ({ [field.id]: field.showInList }) )
//   )
// }


// // Filters the documents array for the one with the id
// export const getRxDocumentById = (docs: RxDocument<Taxonomie, any>[], id: string) => {
//   if (!docs.length)
//     throw new LodgerError('Empty docs provided: %%')
//   const doc = docs.filter(doc => doc._id === id)[0]
//   if (!(doc && isRxDocument(doc)))
//     throw new LodgerError('No document found %%', { id })
//   return doc
// }

//
  /**
   * Active document for taxonomy
  */
//  protected set _activeDocument (docHolder: ActiveDocumentHolder) {
//   let { taxonomie, doc } = docHolder
//   const debug = Debug('lodger:_activeDocument')
//   const gName = `${taxonomie}/activeDoc`
//   const { store } = this

//   if (!store.getters.hasOwnProperty(gName)) {
//     Object.defineProperty(store.getters, gName, {
//       configurable: false,
//       get () { return doc },
//       set (newDoc) { doc = newDoc }
//     })
//   } else {
//     store.getters[gName] = doc
//   }
// }
