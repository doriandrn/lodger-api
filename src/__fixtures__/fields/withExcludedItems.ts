import fields from './normal'

const fieldsWithExcludedItems = [
  ...fields,
  { id: 'x4', excludeFrom: 'db' },
  { id: 'x5', excludeFrom: 'db', required: true }
]

export default fieldsWithExcludedItems
