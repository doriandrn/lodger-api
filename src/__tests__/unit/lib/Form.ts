import { Form, LodgerFormCreator } from '~/lib/Form'

import fields from '@/__fixtures__/forms/fields/normal'
import fieldsWithExcludedItems from '@/__fixtures__/forms/fields/withExcludedItems'

/**
 * DO NOT CHANGE ANY OF THESE
 * as existing tests run on them
 *
 * Rather, create new / extend the stubs / vars and export them
 */
const name = 'xx'
const plural = 'xxs'

const methods = {
  async lol () {},
  syncMethod () {}
}

const stub1: LodgerFormCreator = {
  name,
  plural,
  fields,
  methods
}

const stub2: LodgerFormCreator = {
  name,
  fields: fieldsWithExcludedItems,
  plural
}


describe('Form', () => {
  let __stub1__: LodgerFormConstructor
  let __stub2__: LodgerFormConstructor

  beforeAll(() => {
    __stub1__ = new Form(stub1)
    __stub2__ = new Form(stub2)
    // console.error(__stub2__)
  })


  describe('constructor()', () => {
    describe('negative', () => {
      const name = 'aFormName'
      const plural = 'formsCollection'
      test('throws if fields is empty', () => {
        const fields: Fields = []
        expect(() => new Form({ fields }, name, plural)).toThrow()
      })
    })

  })

  describe('indexable fields', () => {
    // describe('positive', () => {
    //   test('contains all fields\' ids with index: true', () => {
    //     expect(__stub1__.indexables).toContainEqual(

    //     )
    //   })

    //   test('contains keys from commmon methods', () => {

    //   })
    // })
  })

  describe('.value()', () => {
    describe('positive', () => {
      // nu e  cel mai ok test, redo
      test('it contains all keys, even excluded ones', () => {
        expect(Object.keys(__stub1__.value()))
        .toContainEqual(Object.keys(__stub1__.collection.schema.properties))
      })
    })

    describe('negative', () => {

    })
  })

  describe('.plural getter', () => {
    test('returns the collection name', () => {
      expect(__stub1__.plural).toEqual(__stub1__.collection.name)
    })
  })

  describe('.collection', () => {
    describe('positive', () => {
      test('makes collection', () => {
        const { collection } = __stub1__
        expect(collection).toBeDefined()
        expect(collection.schema).toBeDefined()
      })

      test('methods are passed in if existing', () => {
        const { collection } = __stub1__
        expect(collection.methods).toEqual(stub1.methods)
      })

      test('statics are passed', () => {
        const { collection } = __stub1__
        expect(collection.statics).toEqual(stub1.statics)
      })
    })

    describe('schema', () => {
      test('is defined', () => {
        const { schema } = __stub1__.collection
        expect(schema).toBeDefined()
      })

      test('title matches name', () => {
        const { schema } = __stub1__.collection
        expect(schema.title).toBe(name)
      })

      // test('matches fields length', () => {
      //   const { schema } = __stub1__.collection
      //   const keys = Object.keys(schema.properties)
      //   expect(keys.length).toBe(fields.length)
      // })

      describe('.required[]', () => {

        test('adds required fields', () => {
          const { schema } = __stub2__.collection
          const { required } = schema
          expect(required).toContain('x2')
          expect(required).not.toContain('x5')
        })

        test('does NOT contain duplicates', () => {
          const { schema } = __stub2__.collection
          const { required } = schema
          expect(required).toEqual(['x2'])
        })

        test('excluded from db fields dont show up', () => {
          const { schema } = __stub2__.collection
          expect(schema.properties.x5).toBeUndefined()
        })
      })

      test('excludes fields', () => {
        const { schema: { properties } } = __stub2__.collection
        expect(properties).not.toHaveProperty('x5')
      })

      test('indexable fields have -index- prop', () => {
        const { schema: { properties } } = __stub2__.collection
        expect(properties.x2.index).toBeDefined()
      })

      describe('adds common-shared fields', () => {
        test('la - @ - time the doc was added', () => {
          const { schema: { properties } } = __stub2__.collection
          expect(properties.la).toBeDefined()
        })
      })

    })
  })

  describe('.load() - Loads a form by name', () => {
    let form
    const formToLoadAndTest = 'apartament'
    beforeAll(async () => {
      try {
        form = await Form.load(formToLoadAndTest)
      } catch (e) {
        console.error('wtf', e)
      }
    })

    describe('positive', () => {

      test('returns a fully inited <Form> if found and ok', () => {
        expect(form).toBeDefined()
        expect(form.name).toBeDefined()
        expect(form.collection).toBeDefined()
      })

      test('form name is the same  as requested', () => {
        console.error(form)
        expect(form.name).toBe(formToLoadAndTest)
      })
    })

    describe('negative', () => {
      test('throws if called with anything else than string', async () => {
        try {
          await Form.load('')
        } catch (e) {
          expect(e).toBeDefined()
        }

      })
      test('throws for unknown filenames', async () => {
        try {
          await Form.load('ceva')
        } catch (e) {
          expect(e).toBeDefined()
        }
      })
    })
  })


})
