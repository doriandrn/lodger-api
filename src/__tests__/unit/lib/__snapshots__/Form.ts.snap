// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form constructor matches snapshot: xx 1`] = `
Form {
  "$active": false,
  "_onsubmit": Array [
    [Function],
  ],
  "fields": Object {
    "x1": Field {
      "_type": undefined,
      "default": undefined,
      "key": "x1",
      "preview": undefined,
      "storage": "db",
      "type": "string",
      "value": [Function],
    },
    "x2": Field {
      "_index": true,
      "_type": undefined,
      "default": undefined,
      "key": "x2",
      "preview": undefined,
      "storage": "db",
      "type": "string",
      "v": "required|undefined",
      "value": [Function],
    },
    "x3": Field {
      "_type": "number",
      "default": undefined,
      "key": "x3",
      "preview": undefined,
      "storage": "db",
      "type": "number",
      "value": [Function],
    },
    "x4": Field {
      "_type": undefined,
      "default": undefined,
      "key": "x4",
      "preview": undefined,
      "storage": "db",
      "type": "string",
      "value": [Function],
    },
    "x5": Field {
      "_type": undefined,
      "default": undefined,
      "key": "x5",
      "preview": undefined,
      "storage": "db",
      "type": "string",
      "v": "required|undefined",
      "value": [Function],
    },
  },
  "name": "xx",
  "opts": undefined,
  "plural": "xxi",
  "schema": Schema {
    "indexes": Array [
      "x2",
    ],
    "name": "xx",
    "properties": Object {
      "x1": Object {
        "key": "x1",
        "type": "string",
      },
      "x2": Object {
        "_index": true,
        "key": "x2",
        "type": "string",
      },
      "x3": Object {
        "_type": "number",
        "key": "x3",
        "type": "number",
      },
      "x4": Object {
        "key": "x4",
        "type": "string",
      },
      "x5": Object {
        "key": "x5",
        "type": "string",
      },
    },
    "required": Array [
      "x2",
      "x5",
    ],
    "type": "object",
    "version": 0,
  },
}
`;

exports[`Form options .captureTimestamp matches snapshot: options.captureTimestamp 1`] = `
Form {
  "$active": false,
  "_onsubmit": Array [
    [Function],
  ],
  "fields": Object {
    "createdAt": Field {
      "_index": true,
      "_type": "dateTime",
      "default": undefined,
      "preview": undefined,
      "storage": "db",
      "type": "number",
      "value": [Function],
    },
    "updatedAt": Field {
      "_index": true,
      "_type": "dateTime",
      "default": undefined,
      "preview": undefined,
      "storage": "db",
      "type": "number",
      "value": [Function],
    },
    "x1": Field {
      "_type": undefined,
      "default": undefined,
      "key": "x1",
      "preview": undefined,
      "storage": "db",
      "type": "string",
      "value": [Function],
    },
    "x2": Field {
      "_index": true,
      "_type": undefined,
      "default": undefined,
      "key": "x2",
      "preview": undefined,
      "storage": "db",
      "type": "string",
      "v": "required|undefined",
      "value": [Function],
    },
    "x3": Field {
      "_type": "number",
      "default": undefined,
      "key": "x3",
      "preview": undefined,
      "storage": "db",
      "type": "number",
      "value": [Function],
    },
    "x4": Field {
      "_type": undefined,
      "default": undefined,
      "key": "x4",
      "preview": undefined,
      "storage": "db",
      "type": "string",
      "value": [Function],
    },
    "x5": Field {
      "_type": undefined,
      "default": undefined,
      "key": "x5",
      "preview": undefined,
      "storage": "db",
      "type": "string",
      "v": "required|undefined",
      "value": [Function],
    },
  },
  "name": "xx",
  "opts": Object {
    "captureTimestamp": true,
  },
  "plural": "xxi",
  "schema": Schema {
    "indexes": Array [
      "x2",
      "createdAt",
      "updatedAt",
    ],
    "name": "xx",
    "properties": Object {
      "createdAt": Object {
        "_index": true,
        "_type": "dateTime",
        "type": "number",
      },
      "updatedAt": Object {
        "_index": true,
        "_type": "dateTime",
        "type": "number",
      },
      "x1": Object {
        "key": "x1",
        "type": "string",
      },
      "x2": Object {
        "_index": true,
        "key": "x2",
        "type": "string",
      },
      "x3": Object {
        "_type": "number",
        "key": "x3",
        "type": "number",
      },
      "x4": Object {
        "key": "x4",
        "type": "string",
      },
      "x5": Object {
        "key": "x5",
        "type": "string",
      },
    },
    "required": Array [
      "x2",
      "x5",
    ],
    "type": "object",
    "version": 0,
  },
}
`;
